### ISSUE TITLE
Token management issues

### ISSUE THREAD

**Issue by tylerlong** [ringcentral/ringcentral-csharp#43]
First of all, there is a typo in the code: https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Auth.cs#L145  We should definitely fix it.  @dariusdamalakas created a PR for this: https://github.com/ringcentral/ringcentral-csharp/pull/41

More comments from @dariusdamalakas 

> The issue comes when subscription renewal loop ([this code](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Subscription/SubscriptionService.cs#L105)) acquires a new token pair if current `access_token` is expired. At the moment there is no way how we can detect when this refresh happens and persist new token pair in our storage. We store auth tokens in the storage, so that the background process would survive restarts. We don't want to lose `refresh_token` and then contact user to ask him to log in again.

**Comment by tylerlong** [comment]
@dariusdamalakas

I understand the issues you encountered and let's work together to figure our a solution.

First of all, [subscription renew](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Subscription/SubscriptionService.cs#L95) and [authorization renew](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Platform.cs#L73) are two different things. The former is for keep the subscription alive, the latter is for keep the RingCentral access_token valid. And the former depends on the latter (If you don't have a valid access_token, you are not able to renew your subscription).

In order to have long running subscription, we need to:
1. Make sure that authorization renew is handled properly. 
1. Make sure that subscription renew is handled properly.

**Comment by dariusdamalakas** [comment]
@tylerlong thanks for elaborating on this. I don't want to be seen as impolite or arrogant, but I believe you are wrong in both **subscription renew** and **authorization renew**. I'll examine each claim one by one. 

 1. >  **subscription renew**
 > I am thinking maybe we can add some code so that subscription renew also takes care of authentication renew.

 It already does that. Here is the code flow which triggers authentication renew:
  1. SubscriptionService.[Renew](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Subscription/SubscriptionService.cs#L95) called 
  1. Tries to issue a `put` request to update subscription [platform.Put(request)](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Subscription/SubscriptionService.cs#L105)
  1. `Platform.put` method redirects to private [Send](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Platform.cs#L140) method
  1. `Platform.Send` checks if still logged in: [if (!LoggedIn())](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Platform.cs#L145)
  1. `Platform.LoggedIn` does following:
      1. Checks if `access_token` is still valid: [ if (Auth.IsAccessTokenValid())](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Platform.cs#L217)
      1. If `access_token` is not valid, (which is the scenario we are examining here), it checks if `refresh_token` is still valid: [if (Auth.IsRefreshTokenValid())](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Platform.cs#L222)
      1. `refresh_token`s are valid for 7 days by default, so if that is still valid, it locks the object and calls [Refresh();](https://github.com/ringcentral/ringcentral-csharp/blob/master/RingCentral/Platform/Platform.cs#L227)
      1. `Platform.Refresh()` uses `refresh_token` to acquire new token pair, and stores that in `Auth` object by calling `SetData`

So we can see that a subscription renew does indeed flow from sending a put request to checking if refresh token is valid, and then using refresh token to acquire new token pairs.  




 1.  > **authorization renew**
 > It's done manually. If you don't call Refresh from time to time, you access_token will be expired.

I don't believe this is so. Each `Put`, `Delete`, `Post`, `Get` methods goes via `Send` method, which i've shown above actually does call `Refresh`.

To be fair, i'd rename `LoggedIn` method to `EnsureAccessTokenStillValid`, or `IsAccessTokenValidOrRefresh`. Surely, this method is too verbose, but it actually has 2 responsibilities - checking token validity, and refreshing tokens.


Let me know if I got this wrong in any place.

**Comment by tylerlong** [comment]
@dariusdamalakas 

You are right, subscription renew will call `Put` which will call `Refresh` indirectly.

For **manually**, I mean you need to trigger it in some way(directly or indirectly). If your app is running but no activity(no GET/PUT/POST/DELETE), access_token will expire.  Here **manually** means there is no background job to do it (like the subscription renew).

**Comment by dariusdamalakas** [comment]
@tylerlong , ok, let's call this `semi-manual` then, because if we have a subscription renew process running, that means auth refresh will be called every time the subscription is expired. Ok, we are clear on this.

So the issue I am facing is that we need to persist the `refresh_token` and `access_token` to some storage so that if the service crashed we do not lose `refresh_token`. Currently, there is no way how we can hook-in into auth refresh and detect/persist updated tokens. 

Perhaps a simple `platform.AuthDataRefreshed` event would suffice here. This we'd be able to subscribe for the event, and store auth data anywhere we like as soon as it is refreshed.

**Comment by tylerlong** [comment]
I think your request is reasonable.

This project is implemented according to [SDK Reference](https://github.com/ringcentral/sdk-reference/blob/master/platform/Platform.ts). There is no such event in the SDK Reference. But I am willing to add it since it is quite reasonable. Let me quickly do something in dev branch.

**Comment by dariusdamalakas** [comment]
@tylerlong , cool thanks. I'd appreciate if anyone would validate our solution as well here. Our final setup is to have several background processes subscribing to all extensions on multiple ring central accounts. We treat each backend service instance equaly, so each instance subscribes to all accounts and all extensions. I don't feel particularly strong that his is the best scaling strategy, but that is what is on the table at the moment. Does this sound reasonable? Each instance will share the same auth data (loaded from data store). Of course, there is a challenge to correctly refresh token here, i.e. refresh only once, but all other instances fetch refreshed tokens from the store.

**Comment by tylerlong** [comment]
I am afraid you will have permission issues. As you told me that the access_token was granted by user via 3-legged authorization. Does the user have the permission to receive events from all extensions?(it is possible if he is a super user). Does the user have the permission to receive events from other accounts? (He must be a super super user if he does).

**Comment by dariusdamalakas** [comment]
@tylerlong it will be an admin RC user. Otherwise yes, you are right, it would not be able to subscribe to all extensions. I think you got it right - a console app installed as a windows service, fetches tokens from the store once they become available.

**Comment by tylerlong** [comment]
@dariusdamalakas  Please check this commit: https://github.com/ringcentral/ringcentral-csharp/commit/145cc9c5d9ec357fa568d7b9484a7f3d54dfecf2

It's in `dev` branch. And it's not finalized yet.It's tiime for me to go home so I just created a quick and dirty version. Any comments and improvements are welcome.

I created an event, the sender is the `Request`, the args contains the `ApiResponse`. You can get AuthData out of `ApiResponse`. 

It's be merged into master (in the next couple of hours). So maybe you need to compile the dev branch yourself. PRs are welcome if you have better implementations!

**Comment by dariusdamalakas** [comment]
@tylerlong cool, much appreciated. I'll try this out, and let you know how well this is working.

**Comment by LingForCC** [comment]
> We treat each backend service instance equaly, so each instance subscribes to all accounts and all extensions.

@dariusdamalakas Just curious about what application scenario it is. Could you elaborate a bit more regarding the background? It would be nice to improve the SDK with real application scenario. 

Thanks in advance!

**Comment by dariusdamalakas** [comment]
@LingForCC  sure, i've dropped you an email to your gmail account.

**Comment by dariusdamalakas** [comment]
Actually, `PubNub` is the wrong tool for this type of integration that we are doing. Just good plain old http polling will scale much better here instead of real-time subscriptions. 

@tylerlong hold off with `AuthDataRefreshed`.  We have forked your repo and publish an internal nuget package, so we'll experiment and play around with your change, and potentially will come up with suggestions what to improve. 

Thanks all for the input, have a nice weekend.

**Comment by tylerlong** [comment]
`AuthDataRefreshed` is included in version `1.0.1`. Close it for now. Feel free to reopen it if necessary.
